src/content/post/2021-06-14-prototype.md

```diff
@@ -0,0 +1,147 @@
+
+### `Object.prototype`
+
+JS에서 모든 Prototype Chain의 종착점에는 `Object.prototype`이라는 객체가 자리하고 있다. 이 객체는 모든 [Built-in 객체](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects)들의 원형이며, Built-in 객체들의 다양한 공통 특성들을 한 데 가진다. ('프로토타입'이라는 단어를 사용한 이유를 여기서 짐작해볼 수 있겠다.)
+
+<img src="../img/object_prototype.png" alt="object prototype" />
+<br />
+
+`Object.prototype`이라는 하나의 원형에서 출발하여 객체마다 필요에 의하여 기능들을 확장해 나간다. 그러한 과정에서 공통된 특성들은 하나의 객체에 모아 그것들을 참조하게 하여 메모리 자원을 보다 효율적으로 사용할 수 있다. 또한, 특정 객체에서 파생된 인스턴스 객체들에 프로퍼티를 추가하고 싶을 때 하나의 객체에만 선언함으로써 리팩토링의 강점을 취할 수도 있다.
+
+### Prototype Object
+
+그렇다면 아까부터 얘기한 프로토타입 객체는 도대체 뭐 하는 녀석일까? 프로토타입 객체란 공통된 특성(공통된 정도는 쓰임새에 따라 다를 수 있다.)을 상속하기 위한 템플릿 객체이며, 생성자 역할을 할 수 있는 모든 함수에 내장된 객체다. 그렇기 때문에 함수의 `prototype` 프로퍼티를 통해 해당 객체에 접근과 객체 프로퍼티의 수정이 가능하다.
+
+```js
+function construct() {}
+
+console.log(construct.prototype);
+// {constructor: construct(), __proto__: Object}
+
+construct.prototype.a = 2;
+
+console.log(construct.prototype);
+// {a: 2, constructor: construct(), __proto__: Object}
+```
+
+프로토타입 객체의 `constructor`라는 프로퍼티는 해당 객체의 생성자 함수를 가리킨다. 또한 이 객체 역시 `__proto__`를 가지고 있는 걸로 보아, 프로토타입 객체는 원형의 확장이면서도 그 자체로 원형이 될 수 있는 상대적 개념임을 짐작할 수 있다.
+
+## Prototype Based- vs. Class Based-
+
+### Class in ES6
+
+JS는 객체 지향 언어지만 기존의 클래스 기반 프로그래밍 방식을 따르지 않으며 [프로토타입 기반 프로그래밍 방식](https://ko.wikipedia.org/wiki/%ED%94%84%EB%A1%9C%ED%86%A0%ED%83%80%EC%9E%85_%EA%B8%B0%EB%B0%98_%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D)으로 동작한다. 하지만 ES6 이후로 JS 내에서도 클래스 문법이 출현하였는데, 잠시 두 코드가 메커니즘적으로 어떤 차이를 가질지 고민해보자.
+
+```js
+function Person(name) {
+  this.name = name;
+}
+
+Person.prototype.say = function () {
+  console.log(`Hi, My name is ${this.name}.`);
+};
+
+const crew = new Person('Kyle');
+
+console.log(crew.say()); // Hi, My name is Kyle.
+```
+
+```js
+class Person {
+  constructor(name) {
+    this.name = name;
+  }
+
+  say() {
+    console.log(`Hi, My name is ${this.name}.`);
+  }
+}
+
+const crew = new Person('Kyle');
+
+console.log(crew.say()); // Hi, My name is Kyle.
+```
+
+두 코드의 결과는 동일하며, 동작 방식도 거의 차이가 없다. 즉 ES6 클래스 문법은 단지 생성자-프로토타입 패턴을 숨기기 위한 Syntactic Sugar에 불과하다. JS는 여전히 프로토타입 기반 언어이다.
+
+그렇다면 클래스 기반 언어와 JS에서 다루는 클래스의 동작 방식은 구체적으로 어떻게 다른 걸까? (모든 프로토타입 기반 언어에서 클래스를 가지고 있는 것은 아니기에 JS로 한정하여 비교해보려 한다.)
+
+클래스 기반 언어에서 클래스란 **'객체의 기능과 만듦새를 정의한 틀'**이며 생성될 객체에 대한 일종의 청사진으로 이해할 수 있다. 그리고 이것의 인스턴스 화를 통해 생성된 객체를 인스턴스라고 한다. 여기서 인스턴스는 클래스에서 정의된 모든 특성에 대한 복사본(copy)이기 때문에 두 개념은 개별적이며 서로 간의 참조는 존재하지 않는다.
+
+반면에 프로토타입 기반 언어에서는 클래스 - 인스턴스의 개념이 아닌 오직 객체 개념만 존재한다. 다시 말하지만, ES6에 도입된 클래스 문법은 실제 클래스 패턴에 대한 모방일 뿐이다. 이것은 JS의 생성자 함수 패턴을 좀 더 클래스답게 보이기 위한 장치이며 생성된 객체는 함수의 프로토타입 객체에 연결되어 있다.
+
+```js
+class Person {}
+
+const kyle = new Person();
+
+console.log(kyle.__proto__);
+// {constructor: class Person, __proto__: Object}
+```
+
+도입부에서 프로토타입 체계에서 '상속'이라는 표현은 적절치 않을 수 있다고 이야기했는데, 클래스 기반 언어의 관점으로 보았을 때 상속의 동작은 결국 인스턴스 화와 마찬가지로 copy에 기반한 동작이기 때문이다. JS에서 객체들이 동작하는 방식을 고려해본다면 상속보다는 오히려 링크를 통한 위임으로 바라보는 것이 더 적절하지 않을까.
```

Comment:
위임(delegation)이라는 단어가 프로그래밍에서 쓰이는 늬앙스를 생각해보면 조금 적절하지 않을 수도 있을 것 같아요. 참조 정도가 적절한 단어가 되지 않을까 생각하는데 어떻게 생각하시나요?

Reply:
> For a variety of reasons, not the least of which is terminology precedent, "inheritance" (and "prototypal inheritance") and all the other OO terms just do not make sense when considering how JavaScript actually works (not just applied to our forced mental models). Instead, "delegation" is a more appropriate term, because these relationships are not copies but delegation links.

YDKJSY 마지막 요약 글의 한 단락입니다. 저도 이 저자 말이 무조건 맞다! 라는 입장은 아니지만, 해당 파트를 처음부터 읽어보고 이미지를 통한 멘탈 모델을 구체화하였을 때, 이 표현이 적절하다고 느꼈기 때문에 사용하였습니다 :) 사실 멘탈 모델이라는 게 여러 관점에 따라 나누어질 수 있는 것이기 때문에, 이러한 시각도 존재하구나 정도로 받아들여주시면 감사하겠습니다 ☺️

Reply:
위임이 참조 행위를 추상화한 개념으로 읽히네요!
어떤 의도인지 알 수 있었습니다